### PR TITLE
feat: avoid decode barrier when passing between different compute nodes

### LINKED ISSUE #13774: bug: reschedule has much larger memory footprint than expected
### Describe the bug

I've encountered several OOM when running a `risectl scale vertical` with lots of fragments included.
For example,  `risectl scale vertical --fragments 1,2,3,4,5,6,7,8` can result in compute node OOM and subsequent reschedule failure. However `risectl scale vertical --fragments 1,2,3,4` and then `risectl scale vertical --fragments 5,6,7,8` can succeed.

I think the only additional memory required is for those new actors not started yet, which should be small in size. And as the source is paused during rescheduling, the memory footprint of old actors should remain stable. Also in my cases there are no source executors involved in rescheduling.

Here are some numbers in my case:
A compute node uses 100GB memory before rescheduling. During rescheduling it increases to 200GB unitl OOM.

### Error message/log

_No response_

### To Reproduce

_No response_

### Expected behavior

_No response_

### How did you deploy RisingWave?

_No response_

### The version of RisingWave

_No response_

### Additional context

_No response_

### LINKED ISSUE #14533: feat: reduce stream barrier body and only send one copy of it to each CN
When the parallelism in the cluster is relatively high and there are many multi-way join streaming jobs, the barrier body `BarrierMutation` on the stream will be amplified multiple times. When flowing between different compute nodes, it will cause a significant amount of memory usage for prost message decoding and may result in OOM. Here is a solution to fix it, details described as bellow: 

> Some thoughts discussed with @st1page , there is one feasible optimization solution to change the process of the barrier:
> Before sending the barrier, it can be sent to the local barrier manager on compute node first. When injecting the barrier, we can provide the id (`epoch`) only and let the actors to read specific mutation information from local barrier manager if necessary.  By this way, `BarrierMutation` only needs to be decoded once on each compute node.
> 
> _Originally posted by @yezizp2012 in https://github.com/risingwavelabs/risingwave/issues/13060#issuecomment-1876766151_

### PR DESCRIPTION
**Pull request by yezizp2012**
I hereby agree to the terms of the [RisingWave Labs, Inc. Contributor License Agreement](https://gist.github.com/TennyZhuang/f00be7f16996ea48effb049aa7be4d66#file-rw_cla).

## What's changed and what's your intention?

Resolve #14533  Close #13774

In this PR, when receiving the `inject_barrier` request, the compute node will cache the mutation body of the barrier. When an actor reads a barrier from a remote input, it will read from the local cache and fill it in to avoid decoding the mutation body. Similarly, when the dispatcher of a compute node sends a barrier to other nodes, it will erase the mutation. When an actor receives a barrier from another node, it will await until receiving the `inject_barrier` request to read the mutation body and then let the barrier flow.

## Checklist

- [ ] I have written necessary rustdoc comments
- [ ] I have added necessary unit tests and integration tests
- [ ] I have added test labels as necessary. See [details](https://github.com/risingwavelabs/risingwave/blob/main/docs/developer-guide.md#ci-labels-guide).
- [ ] I have added fuzzing tests or opened an issue to track them. (Optional, recommended for new SQL features #7934).
- [ ] My PR contains breaking changes. (If it deprecates some features, please create a tracking issue to remove them in the future).
- [ ] All checks passed in `./risedev check` (or alias, `./risedev c`)
- [ ] My PR changes performance-critical code. (Please run macro/micro-benchmarks and show the results.)

- [ ] My PR contains critical fixes that are necessary to be merged into the latest release. (Please check out the [details](https://github.com/risingwavelabs/risingwave/blob/main/CONTRIBUTING.md))

## Documentation

- [ ] My PR needs documentation updates. (Please use the **Release note** section below to summarize the impact on users)

## Release note

If this PR includes changes that directly affect users or other significant modifications relevant to the community, kindly draft a release note to provide a concise summary of these changes. Please prioritize highlighting the impact these changes will have on users.

</details>

### CODE REVIEW

File: src/stream/src/task/barrier_manager/managed_state.rs [hunk 1]
```diff
@@ -561,88 +579,4 @@ mod tests {
         );
         assert!(managed_barrier_state.epoch_barrier_state_map.is_empty());
     }
-
-    #[tokio::test]
-    async fn test_managed_state_issued_after_collect() {
```
Comment (yezizp2012) [inline src/stream/src/task/barrier_manager/managed_state.rs:566]: We ensured that collect will be called after issued, so the test is no longer needed.

File: src/stream/src/task/barrier_manager/managed_state.rs [hunk 2]
```diff
@@ -287,16 +326,14 @@ impl ManagedBarrierState {
...
-            }) => {
-                let new = collected_actors.insert(actor_id);
-                assert!(new);
-                assert_eq!(curr_epoch, barrier.epoch.curr);
+            })
+            | None => {
+                panic!(
+                    "cannot collect new actor barrier {:?} at current state: Stashed or None",
+                    barrier.epoch,
+                )
```
Comment (BugenZhao) [inline src/stream/src/task/barrier_manager/managed_state.rs:336]: Am I understanding correctly: if the state is now `Stashed`, then it's impossible for an actor to call `collect` as the barrier message must be blocked at the `RemoteInput`.

Can we leave a comment here?

File: src/stream/src/executor/exchange/input.rs [hunk 3]
```diff
@@ -190,7 +195,18 @@ impl RemoteInput {
                             .context("RemoteInput backward permits channel closed.")?;
                     }
 
-                    let msg = msg_res.context("RemoteInput decode message error")?;
+                    let mut msg = msg_res.context("RemoteInput decode message error")?;
+
+                    // Read barrier mutation from local barrier manager and attach it to the barrier message.
+                    if cfg!(not(test)) {
```
Comment (BugenZhao) [inline src/stream/src/executor/exchange/input.rs:201]: Can we assert the `mutation` field to be `None` under `cfg(not(test))`?

**Review by BugenZhao — APPROVED** [review]
LGTM!

File: src/stream/src/task/barrier_manager/managed_state.rs (hunk 2 above)
Comment (yezizp2012) [reply src/stream/src/task/barrier_manager/managed_state.rs:336]: Yes exactly. Let me add some comments here.

File: src/stream/src/task/barrier_manager/tests.rs [hunk 4]
```diff
@@ -171,8 +171,8 @@ async fn test_managed_barrier_collection_before_send_request() -> StreamResult<(
     let barrier = Barrier::new_test_barrier(curr_epoch);
     let epoch = barrier.epoch.prev;
 
-    // Collect a barrier before sending
-    manager.collect(extra_actor_id, &barrier);
+    // Read the mutation after receiving the barrier from remote input.
+    let mutation_reader = pin!(manager.read_barrier_mutation(&barrier));
```
Comment (wenym1) [inline src/stream/src/task/barrier_manager/tests.rs:175]: Might poll the future with `poll_fn` and assert the the future is pending like in other places in this test.

File: src/stream/src/task/barrier_manager/managed_state.rs [hunk 5]
```diff
@@ -71,6 +71,7 @@ pub(super) struct BarrierState {
     curr_epoch: u64,
     inner: ManagedBarrierStateInner,
     kind: BarrierKind,
+    mutation: Option<Arc<Mutation>>,
```
Comment (wenym1) [inline src/stream/src/task/barrier_manager/managed_state.rs:74]: How about adding it to the `Issued` state?

File: src/stream/src/task/barrier_manager.rs [hunk 6]
```diff
@@ -157,6 +157,10 @@ pub(super) enum LocalBarrierEvent {
         actor: ActorId,
         state: BackfillState,
     },
+    ReadBarrierMutation {
```
Comment (wenym1) [inline src/stream/src/task/barrier_manager.rs:160]: May rename the event to `BarrierReceived`. In this future we can use it to monitor the inflight barriers in the actor.

**Review by wenym1 — COMMENTED** [review]
Rest LGTM!

File: src/stream/src/task/barrier_manager.rs (hunk 6 above)
Comment (yezizp2012) [reply src/stream/src/task/barrier_manager.rs:160]: Actually the `ReadBarrierMutation` event is issued by actor, which is trying to read mutation from local cache. `BarrierReceived` sounds a little bit confusing here. 🤔

**Review by wenym1 — APPROVED** [review]

**Pull request merged**